go/vt/vtgate/engine/insert_select.go

```diff
@@ -0,0 +1,419 @@
+	}
+
+	rss, indexesPerRss, err := vcursor.ResolveDestinations(ctx, ins.Keyspace.Name, indexes, destinations)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	queries := make([]*querypb.BoundQuery, len(rss))
+	for i := range rss {
+		bvs := sqltypes.CopyBindVariables(bindVars) // we don't want to create one huge bindvars for all values
+		var mids sqlparser.Values
+		for _, indexValue := range indexesPerRss[i] {
+			index, _ := strconv.Atoi(string(indexValue.Value))
+			if keyspaceIDs[index] != nil {
+				row := sqlparser.ValTuple{}
+				for colOffset, value := range rows[index] {
+					bvName := insertVarOffset(index, colOffset)
+					bvs[bvName] = sqltypes.ValueBindVariable(value)
+					row = append(row, sqlparser.NewArgument(bvName))
+				}
+				mids = append(mids, row)
+			}
+		}
+		rewritten := ins.Prefix + sqlparser.String(mids) + ins.Suffix
+		queries[i] = &querypb.BoundQuery{
+			Sql:           rewritten,
+			BindVariables: bvs,
+		}
+	}
+
+	return rss, queries, nil
+}
+
+func (ins *InsertSelect) execInsertSharded(ctx context.Context, vcursor VCursor, bindVars map[string]*querypb.BindVariable) (*sqltypes.Result, error) {
+	result, err := ins.execSelect(ctx, vcursor, bindVars)
+	if err != nil {
+		return nil, err
+	}
+	if len(result.rows) == 0 {
+		return &sqltypes.Result{}, nil
+	}
+
+	return ins.insertIntoShardedTable(ctx, vcursor, bindVars, result)
+}
+
+func (ins *InsertSelect) description() PrimitiveDescription {
+	other := ins.commonDesc()
+	other["TableName"] = ins.GetTableName()
+
+	if len(ins.VindexValueOffset) > 0 {
+		valuesOffsets := map[string]string{}
+		for idx, ints := range ins.VindexValueOffset {
+			if len(ins.ColVindexes) < idx {
+				panic("ins.ColVindexes and ins.VindexValueOffset do not line up")
+			}
+			vindex := ins.ColVindexes[idx]
+			marshal, _ := json.Marshal(ints)
+			valuesOffsets[vindex.Name] = string(marshal)
+		}
+		other["VindexOffsetFromSelect"] = valuesOffsets
+	}
+
+	return PrimitiveDescription{
+		OperatorType:     "Insert",
+		Keyspace:         ins.Keyspace,
+		Variant:          "Select",
+		TargetTabletType: topodatapb.TabletType_PRIMARY,
+		Other:            other,
+	}
+}
+
+func (ic *InsertCommon) commonDesc() map[string]any {
+	other := map[string]any{
+		"MultiShardAutocommit": ic.MultiShardAutocommit,
+		"QueryTimeout":         ic.QueryTimeout,
+		"InsertIgnore":         ic.Ignore,
+		"InputAsNonStreaming":  ic.ForceNonStreaming,
+		"NoAutoCommit":         ic.PreventAutoCommit,
+	}
+
+	if ic.Generate != nil {
+		if ic.Generate.Values == nil {
+			other["AutoIncrement"] = fmt.Sprintf("%s:Offset(%d)", ic.Generate.Query, ic.Generate.Offset)
+		} else {
+			other["AutoIncrement"] = fmt.Sprintf("%s:Values::%s", ic.Generate.Query, sqlparser.String(ic.Generate.Values))
+		}
+	}
+	return other
+}
+
+func insertVarOffset(rowNum, colOffset int) string {
+	return fmt.Sprintf("_c%d_%d", rowNum, colOffset)
+}
+
+type insertRowsResult struct {
+	rows     []sqltypes.Row
+	insertID int64
```

Comment:
in `sqltypes.Result`, insertID is a `uint64`. I think we should use the same type here

Reply:
in the tabletserver, sequences uses int64 type values and return the result in that format. 
We would need to change that first. Currently not in scope for this PR.